### PR TITLE
Add a script for comparing lint output on branches/PRs

### PR DESCRIPTION
**Pull request by MichaelChirico**
Basically extending @russHyde / @AshesITR script here:

https://gist.github.com/russHyde/568fd5af558c860d97e1b932ca773ff7

And tracking it here for transparency/easy versioning. Put in .dev which can hold various helpers for doing dev on the package.

Have been testing it on some sample CRAN packages but it's tough to get started. Lots of Depends to install from packages chosen at random... library is bloating quickly :smile_cat:

### CODE REVIEW

**Review by russHyde — CHANGES_REQUESTED** [review]
Thanks for adding this, mostly looks sound.
I hadn't heard of gert before, so thanks for that tip.
Would it be possible to allow the user to turn off running the load_all step; and could we use optparse for option handling, rather than rolling our own handler, please.

File: .dev/compare_branches.R [hunk 1]
```diff
@@ -0,0 +1,193 @@
+#!/usr/local/bin/Rscript
+
+# compare the lints obtained before/after a given PR
+# the PR is always compared against master
+
+# arguments
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply .dev/compare_branches.R]: yep, done! i don't have a go-to package for this in R so i usually just make it myself :sweat_smile:

File: .dev/compare_branches.R [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  pr <- 709L
+}
+
+# prioritize packages
+if ("packages" %in% names(args)) {
+  packages <- strsplit(args["packages"], ",", fixed = TRUE)[[1L]]
+} else if ("pkg_dir" %in% names(args)) {
+  packages <- list.files(normalizePath(args["pkg_dir"]), full.names = TRUE)
+} else {
+  packages <- file.path("~", "proj", "code_as_data", "data", "packages")
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply .dev/compare_branches.R]: sure; done

**Comment by MichaelChirico** [comment]
Didn't add the option for turning `load_all` on/off (it's easy enough), i got confused about it... I think we have to run `load_all` to source the `lintr` functions for each branch.

What I think should be optional is running `load_all` for the package sources, which I don't think is happening now? But then I don't know why I ran into some "package not found" issues... I guess it's because `object_usage_linter` does run it for the package code? In which case I'm not sure how to turn it off.

Anyway I'll take a closer look at it later.

**Comment by russHyde** [comment]
Sorry about that. I confused myself. Please disregard my note about devtools::load_all; I mistakenly thought you were loading the analysed package prior to linting.

**Review by russHyde — APPROVED** [review]
LGTM

**Comment by russHyde** [comment]
The failing tests in CI aren't related to this PR.

**Comment by AshesITR** [comment]
Running `load_all()` on the linted package is necessary for `object_usage_linter` (AFAIK the only one needing to know global package state). So we might want to add this in the future.

But I ran into very messy issues while running the tests for packages that are needed by the script and contain binaries (e.g. purrr, rlang, dplyr, tidyr, ...) so that feature should probably come with more safety measures like running in separate R processes and maybe installing the active lintr branch into a temporary local library.

**Comment by MichaelChirico** [comment]
Also needed for `object_name_linter` btw... maybe not loading the full package, but I believe `linter:::namespace_imports` won't work if the package's `Depends` are not available... certainly running into that issue now trying to run a comparison for #755

**Comment by MichaelChirico** [comment]
Added some functionality for running interactively, I think it works pretty nicely. Would be happy to merge now, PTAL

File: .dev/compare_branches.R [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+library(optparse)
+library(dplyr)
+library(purrr)
+library(tibble)
+library(usethis)
+library(gert)
+library(devtools)
+
+if(!file.exists("lintr.Rproj")) {
```
Comment (AshesITR) [inline .dev/compare_branches.R:13]: Please fix the lint

File: .dev/compare_branches.R [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    default = file.path("~", sprintf("lintr_compare_branches_%d.csv", as.integer(Sys.time()))),
+    help = "Destination file to which to write the output"
+  )
+)
+
+params <- optparse::parse_args(optparse::OptionParser(option_list = param_list))
+# treat any skipped arguments from the prompt as missing
+if (interactive()) {
+  for (opt in c("branch", "pr", "packages", "pkg_dir", "sample_size")) {
+    if (params[[opt]] == "") params[[opt]] = NULL
```
Comment (AshesITR) [inline .dev/compare_branches.R:71]: lint

File: .dev/compare_branches.R [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+# prioritize "branch"
+is_branch <- FALSE
+if (!is.null(params$branch)) {
+  branch <- params$branch
+  is_branch <- TRUE
+} else if (!is.null(params$pr)) {
+  pr <- params$pr
+} else {
+  message("Please supply a branch (--branch) or a PR number (--pr)")
+  q("no")
```
Comment (AshesITR) [inline .dev/compare_branches.R:86]: Why not `stop()`? `q()` in interactive seems like a bad idea.

File: .dev/compare_branches.R [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+# prioritize packages
+if (!is.null(params$packages)) {
+  packages <- strsplit(params$packages, ",", fixed = TRUE)[[1L]]
+} else if (!is.null(params$pkg_dir)) {
+  packages <- list.files(normalizePath(params$pkg_dir), full.names = TRUE)
+} else {
+  message("Please supply a comma-separated list of packages (--packages) or a directory of packages (--pkg_dir)")
+  q("no")
```
Comment (AshesITR) [inline .dev/compare_branches.R:96]: dito

**Review by AshesITR — CHANGES_REQUESTED** [review]
Looks mostly good to me.

File: .dev/compare_branches.R (hunk 3 above)
Comment (MichaelChirico) [reply .dev/compare_branches.R]: done

File: .dev/compare_branches.R (hunk 5 above)
Comment (MichaelChirico) [reply .dev/compare_branches.R]: agree

**Review by AshesITR — APPROVED** [review]
LGTM, thanks for polishing. This will be very useful.

**Pull request merged**